plugins/synaptics-rmi/fu-synaptics-rmi-ps2-device.c

```diff
@@ -0,0 +1,947 @@
+	for (guint i = 0; ; i++) {
+		guint8 res = 0;
+		g_autoptr(GError) error_local = NULL;
+		if (do_write) {
+			if (!fu_io_channel_write_raw (self->io_channel,
+						      &buf, sizeof(buf),
+						      timeout,
+						      FU_IO_CHANNEL_FLAG_FLUSH_INPUT |
+						      FU_IO_CHANNEL_FLAG_USE_BLOCKING_IO,
+						      error))
+				return FALSE;
+		}
+		do_write = FALSE;
+
+		for (;;) {
+			/* attempt to read acknowledge... */
+			if (!fu_synaptics_rmi_ps2_device_read_ack (self, &res, &error_local)) {
+				if (i > 3) {
+					g_propagate_prefixed_error (error,
+							    g_steal_pointer (&error_local),
+							    "read ack failed: ");
+					return FALSE;
+				}
+				g_warning ("read ack failed: %s, retrying", error_local->message);
+				break;
+			}
+			if (res == edpsAcknowledge)
+				return TRUE;
+			if (res == edpsResend) {
+				do_write = TRUE;
+				g_debug ("resend");
+				g_usleep (G_USEC_PER_SEC);
+				break;
+			}
+			if (res == edpsError) {
+				do_write = TRUE;
+				g_debug ("error");
+				g_usleep (1000 * 10);
+				break;
+			}
+			g_debug ("other response: 0x%x", res);
+			g_usleep (1000 * 10);
+		}
+		if (i >= 3) {
+			g_set_error_literal (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_NOT_SUPPORTED,
+					     "cannot write byte after retries");
+			return FALSE;
+		}
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_rmi_ps2_device_set_resolution_sequence (FuSynapticsRmiPs2Device *self,
+						     guint8 arg,
+						     gboolean send_e6s,
+						     GError **error)
+{
+	/* send set scaling twice if send_e6s */
+	for (gint i = send_e6s ? 2 : 1; i > 0; --i) {
+		if (!fu_synaptics_rmi_ps2_device_write_byte (self, edpAuxSetScaling1To1, 50, error))
+			return FALSE;
+	}
+	for (gint i = 3; i >= 0; --i) {
+		guint8 ucTwoBitArg = (arg >> (i * 2)) & 0x3;
+		if (!fu_synaptics_rmi_ps2_device_write_byte (self, edpAuxSetResolution, 50, error)) {
+			return FALSE;
+		}
+		if (!fu_synaptics_rmi_ps2_device_write_byte (self, ucTwoBitArg, 50, error))
+			return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_rmi_ps2_device_status_request_sequence (FuSynapticsRmiPs2Device *self,
+						     guint8 ucArgument,
+						     guint32 *buf,
+						     GError **error)
+{
+	gboolean success = FALSE;
+
+	/* allow 3 retries */
+	for (guint i = 0; i < 3; ++i) {
+		g_autoptr(GError) error_local = NULL;
+		if (!fu_synaptics_rmi_ps2_device_set_resolution_sequence (self,
+									  ucArgument,
+									  FALSE,
+									  &error_local)) {
+			g_debug ("failed set try #%u: %s",
+				 i, error_local->message);
+			continue;
+		}
+		if (!fu_synaptics_rmi_ps2_device_write_byte (self,
+							     edpAuxStatusRequest,
+							     10,
+							     &error_local)) {
+			g_debug ("failed write try #%u: %s",
+				 i, error_local->message);
+			continue;
+		}
+		success = TRUE;
+		break;
+	}
+	if (success == FALSE) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED, "failed");
```

Comment:
Would it be useful to save the final error from the for loop and propagate it?

Reply:
Well, it can only be a timeout, but sure, G_IO_ERROR_TIMED_OUT is a better error code for sure. Will fix.